torchao/csrc/cpu/paged_attention_kernel.cpp

```diff
@@ -0,0 +1,535 @@
+    vec_tmp_sum += tmp2;
+    tmp2.store(out + i);
+  }
+  tmp_sum = at::vec::vec_reduce_all<T1>(
+      [](at::vec::Vectorized<T1> &x, at::vec::Vectorized<T1> &y) {
+        return x + y;
+      },
+      vec_tmp_sum);
+  for (; i < size; i++) {
+    auto tmp0 = a[i];
+    auto tmp1 = tmp0 - val;
+    auto tmp2 = exp(tmp1);
+    tmp_sum += tmp2;
+    out[i] = tmp2;
+  }
+  val = tmp_sum;
+}
+
+// 1) out = a * scale
+// 2) max = max(out)
+template <typename scalar_t>
+inline void _mul_reduce_max_fusion_kernel(scalar_t *a, const scalar_t &scale,
+                                          const int &size, scalar_t *out,
+                                          scalar_t &max) {
+  auto vec_size = at::vec::Vectorized<scalar_t>::size();
+  auto vec_scale = at::vec::Vectorized<scalar_t>(scale);
+  scalar_t tmp_max = -std::numeric_limits<scalar_t>::infinity();
+  auto vec_tmp_max = at::vec::Vectorized<scalar_t>(tmp_max);
+  long i = 0;
+  for (; i < vec_size * (size / vec_size); i += vec_size) {
+    auto tmp0 = at::vec::Vectorized<scalar_t>::loadu(a + i);
+    auto tmp1 = tmp0 * vec_scale;
+    vec_tmp_max = at::vec::maximum(vec_tmp_max, tmp1);
+    tmp1.store(out + i);
+  }
+  tmp_max = at::vec::vec_reduce_all<scalar_t>(
+      [](at::vec::Vectorized<scalar_t> &x, at::vec::Vectorized<scalar_t> &y) {
+        return at::vec::maximum(x, y);
+      },
+      vec_tmp_max);
+  for (; i < size; i++) {
+    auto tmp0 = a[i];
+    auto tmp1 = tmp0 * scale;
+    tmp_max = std::max(tmp_max, tmp1);
+    out[i] = tmp1;
+  }
+  max = tmp_max;
+}
+
+void reshape_attn_mask_to_4d(at::Tensor &attn_mask, int64_t batchSize,
+                             int64_t num_head, int64_t qSize, int64_t kvSize) {
+  // Support mask shapes:
+  // 2d: ({Q_seq_len, 1}  x {KV_seq_len, 1})
+  // 4d: ({Batch, 1} x {Num_heads, 1} x {Q_seq_len, 1}  x {KV_seq_len, 1})
+  // Guaranteed in check_attn_mask_shape
+  int64_t attn_mask_size_0 = 1;
+  int64_t attn_mask_size_1 = 1;
+  if (attn_mask.dim() == 4) {
+    if (attn_mask.size(0) == batchSize) {
+      attn_mask_size_0 = batchSize;
+    }
+    if (attn_mask.size(1) == num_head) {
+      attn_mask_size_1 = num_head;
+    }
+  }
+  attn_mask = attn_mask
+                  .view({attn_mask_size_0, attn_mask_size_1, attn_mask.size(-2),
+                         attn_mask.size(-1)})
+                  .expand({attn_mask_size_0, attn_mask_size_1, qSize, kvSize});
+}
+
+/**
+ * Performs scale-dot-product for the next token based on cached key-value
+ * attention.
+ *
+ * This function computes the attention weights and applies the attention
+ * mechanism to obtain the final output. It takes in tensors representing the
+ * query, key cache, value cache, head mapping, scale, block tables, context
+ * lengths, block size The
+ * output tensor is updated with the computed attention values.
+ *
+ * @param out           Output tensor [num_seqs, 1, num_heads, head_size].
+ * @param query         Query tensor [num_seqs, 1, num_heads, head_size].
```

Comment:
Please add runtime assertion in the code to make sure the query has seq length 1 here. BTW, can we extend the implementation to support query seq length > 1 which can benefit chunked prefill and multi-turn conversation cases?